hypothesis-python/tests/numpy/test_from_type.py

```diff
@@ -0,0 +1,176 @@
+    arr = from_type(np.ndarray[typ]).example()
+    assert isinstance(arr, np.ndarray)
+    assert arr.dtype.type == typ
+
+    arr = from_type(np.ndarray[typing.Any, typ]).example()
+    assert isinstance(arr, np.ndarray)
+    assert arr.dtype.type == typ
+
+
+@pytest.mark.skipif(
+    NDArray is None,
+    reason="numpy.typing is not available",
+)
+@pytest.mark.parametrize("typ", STANDARD_TYPES_TYPE)
+def test_resolves_specified_NDArray_type(typ):
+    arr = from_type(NDArray[typ]).example()
+    assert isinstance(arr, np.ndarray)
+    assert arr.dtype.type == typ
+
+
+@pytest.mark.skipif(
+    ArrayLike is None,
+    reason="numpy.typing is not available",
+)
+@given(arr_like=from_type(ArrayLike))
+def test_resolves_ArrayLike_type(arr_like):
+    arr = np.array(arr_like)
+    assert isinstance(arr, np.ndarray)
+    # The variation is too large to assert anything else about arr, but the
+    # ArrayLike contract just says that it can be coerced into an array (which
+    # we just did).
+
+
+@pytest.mark.skipif(
+    _NestedSequence is None,
+    reason="numpy._typing is not available",
+)
+def test_resolves_specified_NestedSequence():
+    @given(seq=from_type(_NestedSequence[int]))
+    def test(seq):
+        assert hasattr(seq, "__iter__")
+
+        def flatten(lst):
+            for el in lst:
+                try:
+                    yield from flatten(el)
+                except TypeError:
+                    yield el
+
+        assert all(isinstance(i, int) for i in flatten(seq))
+
+    test()
+
+
+@pytest.mark.skipif(
+    _NestedSequence is None,
+    reason="numpy._typing is not available",
+)
+@given(seq=from_type(_NestedSequence))
+def test_resolves_unspecified_NestedSequence(seq):
+    assert hasattr(seq, "__iter__")
+
+
+@pytest.mark.skipif(
+    _SupportsArray is None,
+    reason="numpy._typing is not available",
+)
+@given(arr=from_type(_SupportsArray))
+def test_resolves_unspecified_SupportsArray(arr):
+    assert hasattr(arr, "__array__")
+
+
+@pytest.mark.skipif(
+    _SupportsArray is None,
+    reason="numpy._typing is not available",
+)
+def test_resolves_SupportsArray():
+    @given(arr=from_type(_SupportsArray[int]))
+    def test(arr):
+        assert hasattr(arr, "__array__")
+        assert np.asarray(arr).dtype.kind == "i"
+
+    test()
+
+
+@pytest.mark.skipif(
+    _NestedSequence is None or _SupportsArray is None,
+    reason="numpy._typing is not available",
+)
+def test_resolve_ArrayLike_equivalent():
+    # This is the current (1.24.3) definition of ArrayLike,
+    # with problematic parts commented out.
+    ArrayLike_like = typing.Union[
+        _SupportsArray,
+        # _NestedSequence[_SupportsArray],
+        bool,
+        int,
+        float,
+        complex,
+        str,
+        bytes,
+        _NestedSequence[
+            typing.Union[
+                bool,
+                int,
+                float,
+                complex,
+                str,
+                # bytes,
```

Comment:
Is this specifically that _bytes doesn't work_, or _we can't mix str and bytes_?  Either is fine, I'm just keen to keep an accurate record of the problems you hit because I expect it to come up later.

Reply:
The latter. `[b'\x80', '']` fails to convert to `ndarray` with some ascii-decode-error. Probably a numpy bug. I added a comment-link to the PR that shows this example, but I can inline it instead

Reply:
Oh dear, this looks pretty nasty: if you pass a mixture of str and bytes, Numpy will try to decode all the bytes as ascii-encoded strings - so that's the minimal non-ascii-decodable bytestring, and the minimal unicode string.

Arguably mixing strings and bytes like this should be an upstream error, but... `np.array(["unicode", 1.337])` also casts everything to unicode strings, so I guess the problem is only the assumption that bytestrings must be ascii-encoded.

How about we just limit our strategy to ascii-compatible bytes, and I'll report this upstream?